File: MIDTERM-LAB-1/productCategoryProfitability.js
/*
 * PROGRAMMING 2 – MACHINE PROBLEM
 * University of Perpetual Help System DALTA – Molino Campus
 * BS Information Technology - Game Development
 * Dataset: https://www.kaggle.com/datasets/asaniczka/video-game-sales-2024
 *
 * productCategoryProfitability.js  (Node.js)
 *
 * ANALYSIS: Product Category Profitability
 *   - Reads CSV from a user-supplied file path
 *   - Groups records by genre (category)
 *   - Computes total sales and average sale per category
 *   - Identifies most & least profitable categories
 *
 * Run: node productCategoryProfitability.js
 */

'use strict';

const fs       = require('fs');
const path     = require('path');
const readline = require('readline');

// ─────────────────────────────────────────────────────────────────────────────
//  Helper: parse a single CSV line (handles double-quoted fields)
// ─────────────────────────────────────────────────────────────────────────────
function parseCSVLine(line) {
    const cols    = [];
    let   inQuote = false;
    let   current = '';

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (ch === '"') {
            inQuote = !inQuote;
        } else if (ch === ',' && !inQuote) {
            cols.push(current.trim());
            current = '';
        } else {
            current += ch;
        }
    }
    cols.push(current.trim());
    return cols;
}

// ─────────────────────────────────────────────────────────────────────────────
//  Module: loadDataset(filePath) → Array of record objects
// ─────────────────────────────────────────────────────────────────────────────
function loadDataset(filePath) {
    const rawContent = fs.readFileSync(filePath, 'utf8');
    const lines      = rawContent.split(/\r?\n/).filter(l => l.trim() !== '');

    if (lines.length < 2) {
        throw new Error('CSV file is empty or contains only a header row.');
    }

    // ── Detect column indices from header ──────────────────────────────────
    const headers = parseCSVLine(lines[0]).map(h => h.toLowerCase().replace(/"/g, ''));

    const titleIdx = headers.findIndex(h => ['title','name','game'].includes(h));
    const genreIdx = headers.findIndex(h => ['genre','category'].includes(h));
    const salesIdx = headers.findIndex(h => ['total_sales','sales','totalsales'].includes(h));

    if (genreIdx === -1 || salesIdx === -1) {
        throw new Error(
            `Required columns not found.\n` +
            `Expected: 'genre' (or 'category') and 'total_sales' (or 'sales').\n` +
            `Found headers: ${lines[0]}`
        );
    }

    // ── Parse data rows ────────────────────────────────────────────────────
    const records = [];
    let   skipped = 0;

    for (let i = 1; i < lines.length; i++) {
        const cols  = parseCSVLine(lines[i]);
        const genre = genreIdx < cols.length ? cols[genreIdx].replace(/"/g, '') : '';
        const rawSales = salesIdx < cols.length ? cols[salesIdx].replace(/[",]/g, '') : '';
        const sales  = parseFloat(rawSales);
        const title  = titleIdx >= 0 && titleIdx < cols.length
                     ? cols[titleIdx].replace(/"/g, '') : 'Unknown';

        if (!genre || isNaN(sales)) { skipped++; continue; }
        records.push({ title, genre, sales });
    }

    if (skipped > 0) console.log(`[INFO]  ${skipped} malformed row(s) skipped.`);
    return records;
}

// ─────────────────────────────────────────────────────────────────────────────
//  Module: analyzeByCategory(records) → sorted category stats
// ─────────────────────────────────────────────────────────────────────────────
function analyzeByCategory(records) {
    const statsMap = {};   // { genre: { total, count } }

    for (const rec of records) {
        if (!statsMap[rec.genre]) statsMap[rec.genre] = { total: 0, count: 0 };
        statsMap[rec.genre].total += rec.sales;
        statsMap[rec.genre].count += 1;
    }

    // Convert to array and sort by total sales descending
    return Object.entries(statsMap)
        .map(([category, { total, count }]) => ({
            category,
            totalSales : Math.round(total * 100) / 100,
            avgSale    : Math.round((total / count) * 100) / 100,
            count
        }))
        .sort((a, b) => b.totalSales - a.totalSales);
}

// ─────────────────────────────────────────────────────────────────────────────
//  Module: displayResults(stats, totalRecords)
// ─────────────────────────────────────────────────────────────────────────────
function displayResults(stats, totalRecords) {
    const fmt  = (n) => n.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 });
    const pad  = (s, w) => String(s).padEnd(w);
    const lpad = (s, w) => String(s).padStart(w);
    const rep  = (ch, n) => ch.repeat(n);

    // Grand total for share %
    const grandTotal = stats.reduce((sum, s) => sum + s.totalSales, 0);

    // ================================================================
    //  BANNER
    // ================================================================
    console.log();
    console.log('  ' + rep('*', 72));
    console.log('  *' + rep(' ', 70) + '*');
    console.log('  *     VIDEO GAME SALES 2024  --  CATEGORY PROFITABILITY REPORT       *');
    console.log('  *     University of Perpetual Help System DALTA - Molino Campus      *');
    console.log('  *     BS Information Technology - Game Development                   *');
    console.log('  *' + rep(' ', 70) + '*');
    console.log('  ' + rep('*', 72));

    // ================================================================
    //  DATASET SUMMARY BOX
    // ================================================================
    console.log();
    console.log('  +--------------------------+--------------------------+');
    console.log(`  |  Total Games Loaded      |  ${lpad(totalRecords.toLocaleString(), 24)}  |`);
    console.log(`  |  Unique Categories       |  ${lpad(stats.length, 24)}  |`);
    console.log(`  |  Grand Total Sales (M)   |  ${lpad(fmt(grandTotal), 24)}  |`);
    console.log('  +--------------------------+--------------------------+');

    // ================================================================
    //  MAIN TABLE
    // ================================================================
    console.log();
    console.log('  ' + rep('=', 78));
    console.log(`  ${'RANK'.padEnd(5)}  |  ${'CATEGORY'.padEnd(18)}  |  ${'TOTAL SALES'.padStart(12)}  |  ${'AVG SALE'.padStart(10)}  |  ${'# GAMES'.padStart(7)}  |  ${'SHARE'.padStart(5)}`);
    console.log('  ' + rep('=', 78));

    stats.forEach((s, i) => {
        const share  = ((s.totalSales / grandTotal) * 100).toFixed(1);
        const marker = i % 2 === 0 ? '>' : ' ';
        const rank   = String(i + 1);

        console.log(`  ${marker}${pad(rank, 4)}  |  ${pad(s.category, 18)}  |  ${lpad(fmt(s.totalSales), 12)}  |  ${lpad(fmt(s.avgSale), 10)}  |  ${lpad(s.count, 7)}  |  ${lpad(share + '%', 5)}`);

        // Divider every 5 rows
        if ((i + 1) % 5 === 0 && (i + 1) < stats.length) {
            console.log('  ' + rep('-', 78));
        }
    });

    console.log('  ' + rep('=', 78));

    // ================================================================
    //  FINDINGS PANEL
    // ================================================================
    const best       = stats[0];
    const worst      = stats[stats.length - 1];
    const bestShare  = ((best.totalSales  / grandTotal) * 100).toFixed(1);
    const worstShare = ((worst.totalSales / grandTotal) * 100).toFixed(1);

    console.log();
    console.log('  +--[ FINDINGS ]' + rep('-', 62) + '+');
    console.log('  |' + rep(' ', 76) + '|');

    const bestLine  = `[#1 MOST PROFITABLE]   ${pad(best.category, 18)}   Sales: ${lpad(fmt(best.totalSales), 12)}   Share: ${lpad(bestShare + '%', 5)}`;
    const worstLine = `[#${stats.length} LEAST PROFITABLE]  ${pad(worst.category, 18)}   Sales: ${lpad(fmt(worst.totalSales), 12)}   Share: ${lpad(worstShare + '%', 5)}`;

    console.log(`  |   ${bestLine.padEnd(73)}|`);
    console.log('  |' + rep(' ', 76) + '|');
    console.log(`  |   ${worstLine.padEnd(73)}|`);
    console.log('  |' + rep(' ', 76) + '|');
    console.log('  +' + rep('-', 76) + '+');
    console.log();
}

// ─────────────────────────────────────────────────────────────────────────────
//  File-path prompt loop (uses readline + fs.existsSync)
// ─────────────────────────────────────────────────────────────────────────────
function askFilePath(rl, callback) {
    rl.question('\nEnter dataset file path: ', (inputPath) => {
        const trimmed = inputPath.trim();

        if (!fs.existsSync(trimmed)) {
            console.log('[ERROR] File does not exist. Please try again.');
            askFilePath(rl, callback);
        } else if (!fs.statSync(trimmed).isFile()) {
            console.log('[ERROR] Path does not point to a file. Please try again.');
            askFilePath(rl, callback);
        } else if (!trimmed.toLowerCase().endsWith('.csv')) {
            console.log('[ERROR] File does not appear to be a CSV. Please try again.');
            askFilePath(rl, callback);
        } else {
            console.log(`[OK]    File found: ${path.resolve(trimmed)}`);
            callback(trimmed);
        }
    });
}

// ─────────────────────────────────────────────────────────────────────────────
//  Main entry point
// ─────────────────────────────────────────────────────────────────────────────
function main() {
    console.log();
    console.log('  +-----------------------------------------------------------------+');
    console.log('  |                                                                 |');
    console.log('  |   PROGRAMMING 2  --  Machine Problem                           |');
    console.log('  |   Product Category Profitability Analyzer                      |');
    console.log('  |   Dataset : Video Game Sales 2024  (Kaggle)                    |');
    console.log('  |                                                                 |');
    console.log('  +-----------------------------------------------------------------+');
    console.log();

    const rl = readline.createInterface({
        input  : process.stdin,
        output : process.stdout
    });

    askFilePath(rl, (filePath) => {
        rl.close();

        try {
            // Step 1 – Load dataset
            console.log('\n[INFO]  Loading dataset...');
            const records = loadDataset(filePath);

            if (records.length === 0) {
                console.log('[ERROR] No valid records found in the file.');
                return;
            }

            console.log(`[INFO]  ${records.length.toLocaleString()} records loaded.`);

            // Step 2 – Analyse
            const stats = analyzeByCategory(records);

            // Step 3 – Display
            displayResults(stats, records.length);

        } catch (err) {
            console.error('[ERROR]', err.message);
        }
    });
}

main();

File: MIDTERM-LAB-1/ProductCategoryProfitability.java
/*
 * PROGRAMMING 2 – MACHINE PROBLEM
 * University of Perpetual Help System DALTA – Molino Campus
 * BS Information Technology - Game Development
 * Dataset: https://www.kaggle.com/datasets/asaniczka/video-game-sales-2024
 *
 * ProductCategoryProfitability.java — Main class
 *
 * ANALYSIS: Product Category Profitability
 *   - Groups records by genre (category)
 *   - Computes total sales and average sale per category
 *   - Identifies most & least profitable categories
 */

import java.io.*;
import java.util.*;

public class ProductCategoryProfitability {

    // ------------------------------------------------------------------ //
    //  File validation & loading                                          //
    // ------------------------------------------------------------------ //

    /**
     * Prompts the user for a CSV file path, validates it, and returns the
     * File object.  Loops until a valid, readable CSV file is provided.
     */
    private static File promptForFile(Scanner input) {
        while (true) {
            System.out.print("\nEnter dataset file path: ");
            String path = input.nextLine().trim();
            File file = new File(path);

            if (!file.exists()) {
                System.out.println("[ERROR] File does not exist. Please try again.");
            } else if (!file.isFile()) {
                System.out.println("[ERROR] Path does not point to a file. Please try again.");
            } else if (!file.canRead()) {
                System.out.println("[ERROR] File is not readable. Please try again.");
            } else if (!path.toLowerCase().endsWith(".csv")) {
                System.out.println("[ERROR] File does not appear to be a CSV. Please try again.");
            } else {
                System.out.println("[OK]    File found: " + file.getAbsolutePath());
                return file;
            }
        }
    }

    /**
     * Reads the CSV file and returns a list of DataRecord objects.
     * Expected columns (case-insensitive header detection):
     *   title, genre, total_sales
     */
    private static List<DataRecord> loadRecords(File file) throws IOException {
        List<DataRecord> records = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String headerLine = br.readLine();
            if (headerLine == null) throw new IOException("CSV file is empty.");

            // Detect column indices from header
            String[] headers = headerLine.split(",");
            int titleIdx = -1, genreIdx = -1, salesIdx = -1;

            for (int i = 0; i < headers.length; i++) {
                String h = headers[i].trim().toLowerCase().replace("\"", "");
                if (h.equals("title") || h.equals("name") || h.equals("game"))          titleIdx = i;
                if (h.equals("genre") || h.equals("category"))                           genreIdx = i;
                if (h.equals("total_sales") || h.equals("sales") || h.equals("totalsales")) salesIdx = i;
            }

            if (genreIdx == -1 || salesIdx == -1) {
                throw new IOException(
                    "Required columns not found in CSV.\n" +
                    "Expected: 'genre' (or 'category') and 'total_sales' (or 'sales').\n" +
                    "Found headers: " + headerLine
                );
            }

            String line;
            int skipped = 0;

            while ((line = br.readLine()) != null) {
                // Handle quoted fields naively (split on comma outside quotes)
                String[] cols = splitCSVLine(line);

                try {
                    String title = (titleIdx >= 0 && titleIdx < cols.length)
                            ? cols[titleIdx].trim().replace("\"", "") : "Unknown";
                    String genre = cols[genreIdx].trim().replace("\"", "");
                    double sales = Double.parseDouble(
                            cols[salesIdx].trim().replace("\"", "").replace(",", ""));

                    if (!genre.isEmpty()) {
                        records.add(new DataRecord(title, genre, sales));
                    }
                } catch (NumberFormatException | ArrayIndexOutOfBoundsException e) {
                    skipped++;   // skip malformed rows silently
                }
            }

            if (skipped > 0) {
                System.out.printf("[INFO]  %d malformed row(s) skipped.%n", skipped);
            }
        }

        return records;
    }

    /** Returns a String of n repeated copies of ch (replaces String.repeat for Java 8+). */
    private static String repeat(char ch, int n) {
        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; i++) sb.append(ch);
        return sb.toString();
    }

    /** Simple CSV line splitter that respects double-quoted fields. */
    private static String[] splitCSVLine(String line) {
        List<String> result = new ArrayList<>();
        boolean inQuotes = false;
        StringBuilder sb = new StringBuilder();
        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                result.add(sb.toString());
                sb.setLength(0);
            } else {
                sb.append(c);
            }
        }
        result.add(sb.toString());
        return result.toArray(new String[0]);
    }

    // ------------------------------------------------------------------ //
    //  Analytics                                                          //
    // ------------------------------------------------------------------ //

    private static void analyzeAndDisplay(List<DataRecord> records) {
        // Accumulators: category -> {totalSales, count}
        Map<String, Double>  totalSalesMap = new LinkedHashMap<>();
        Map<String, Integer> countMap      = new LinkedHashMap<>();

        for (DataRecord r : records) {
            String cat = r.getGenre();
            totalSalesMap.merge(cat, r.getTotalSales(), Double::sum);
            countMap.merge(cat, 1, Integer::sum);
        }

        // Sort by total sales descending
        List<Map.Entry<String, Double>> sorted = new ArrayList<>(totalSalesMap.entrySet());
        sorted.sort((a, b) -> Double.compare(b.getValue(), a.getValue()));

        // Compute grand total for share %
        double grandTotal = 0;
        for (double v : totalSalesMap.values()) grandTotal += v;

        // ================================================================
        //  BANNER
        // ================================================================
        System.out.println();
        System.out.println("  " + repeat('*', 72));
        System.out.println("  *" + repeat(' ', 70) + "*");
        System.out.println("  *     VIDEO GAME SALES 2024  --  CATEGORY PROFITABILITY REPORT" + repeat(' ', 7) + "*");
        System.out.println("  *     University of Perpetual Help System DALTA - Molino Campus" + repeat(' ', 6) + "*");
        System.out.println("  *     BS Information Technology - Game Development" + repeat(' ', 20) + "*");
        System.out.println("  *" + repeat(' ', 70) + "*");
        System.out.println("  " + repeat('*', 72));

        // ================================================================
        //  DATASET SUMMARY BOX
        // ================================================================
        System.out.println();
        System.out.println("  +--------------------------+--------------------------+");
        System.out.printf ("  |  Total Games Loaded      |  %,24d  |%n", records.size());
        System.out.printf ("  |  Unique Categories       |  %24d  |%n", sorted.size());
        System.out.printf ("  |  Grand Total Sales (M)   |  %,24.2f  |%n", grandTotal);
        System.out.println("  +--------------------------+--------------------------+");

        // ================================================================
        //  MAIN TABLE
        // ================================================================
        System.out.println();
        System.out.println("  " + repeat('=', 78));
        System.out.printf ("  %-4s  |  %-18s  |  %12s  |  %10s  |  %7s  |  %5s%n",
                "RANK", "CATEGORY", "TOTAL SALES", "AVG SALE", "# GAMES", "SHARE");
        System.out.println("  " + repeat('=', 78));

        for (int i = 0; i < sorted.size(); i++) {
            String cat    = sorted.get(i).getKey();
            double total  = sorted.get(i).getValue();
            int    count  = countMap.get(cat);
            double avg    = total / count;
            double share  = (total / grandTotal) * 100.0;

            // Alternate row markers for readability
            String marker = (i % 2 == 0) ? ">" : " ";

            System.out.printf("  %s%-3d  |  %-18s  |  %,12.2f  |  %,10.2f  |  %7d  |  %4.1f%%%n",
                    marker, (i + 1), cat, total, avg, count, share);

            // Separator every 5 rows
            if ((i + 1) % 5 == 0 && (i + 1) < sorted.size()) {
                System.out.println("  " + repeat('-', 78));
            }
        }

        System.out.println("  " + repeat('=', 78));

        // ================================================================
        //  FINDINGS PANEL
        // ================================================================
        Map.Entry<String, Double> best  = sorted.get(0);
        Map.Entry<String, Double> worst = sorted.get(sorted.size() - 1);
        double bestShare  = (best.getValue()  / grandTotal) * 100.0;
        double worstShare = (worst.getValue() / grandTotal) * 100.0;

        System.out.println();
        System.out.println("  +--[ FINDINGS ]" + repeat('-', 62) + "+");
        System.out.println("  |" + repeat(' ', 76) + "|");

        String bestLine  = String.format("[#1 MOST PROFITABLE]   %-18s   Sales: %,12.2f   Share: %4.1f%%",
                best.getKey(), best.getValue(), bestShare);
        String worstLine = String.format("[#%d LEAST PROFITABLE]  %-18s   Sales: %,12.2f   Share: %4.1f%%",
                sorted.size(), worst.getKey(), worst.getValue(), worstShare);

        System.out.printf("  |   %-73s|%n", bestLine);
        System.out.println("  |" + repeat(' ', 76) + "|");
        System.out.printf("  |   %-73s|%n", worstLine);
        System.out.println("  |" + repeat(' ', 76) + "|");
        System.out.println("  +" + repeat('-', 76) + "+");
        System.out.println();
    }

    // ------------------------------------------------------------------ //
    //  Entry point                                                        //
    // ------------------------------------------------------------------ //

    public static void main(String[] args) {
        System.out.println();
        System.out.println("  +-----------------------------------------------------------------+");
        System.out.println("  |                                                                 |");
        System.out.println("  |   PROGRAMMING 2  --  Machine Problem                           |");
        System.out.println("  |   Product Category Profitability Analyzer                      |");
        System.out.println("  |   Dataset : Video Game Sales 2024  (Kaggle)                    |");
        System.out.println("  |                                                                 |");
        System.out.println("  +-----------------------------------------------------------------+");
        System.out.println();

        Scanner input = new Scanner(System.in);

        try {
            // Step 1 – Prompt & validate file path
            File file = promptForFile(input);

            // Step 2 – Load dataset
            System.out.println("\n[INFO]  Loading dataset...");
            List<DataRecord> records = loadRecords(file);

            if (records.isEmpty()) {
                System.out.println("[ERROR] No valid records found in the file.");
                return;
            }

            // Step 3 – Analyse & display
            analyzeAndDisplay(records);

        } catch (IOException e) {
            System.out.println("[ERROR] " + e.getMessage());
        } finally {
            input.close();
        }
    }
}
